lib/Mail/Milter/Authentication/Resolver.pm

```diff
@@ -0,0 +1,82 @@
+package Mail::Milter::Authentication::Resolver;
+use strict;
+use warnings;
+# VERSION
+use base 'Net::DNS::Resolver';
+
+=head1 DESCRIPTION
+
+Superclass for Net::DNS::Resolver, Versions of Net::DNS::Resolver from 1.03 up (to at least
+1.18 at time of writing) do not timeout as expected. This introduces a wrapper timeout around
+the query, send, and search calls which will fire 0.1 seconds after the timeout value passed
+to Net::DNS::Resolver
+
+=cut
+
+{
+  sub new {
+    my $class = shift;
+    my %args = @_;
+    my $self = $class->SUPER::new( @_ );
+    $self->{ _handler } = $args{_handler};
+    return $self;
+  }
+}
+
+sub _do {
+  my $self = shift;
+  my $what = shift;
+  my $handler = $self->{_handler};
+  my $config = $handler->config();
+  my $timeout = $config->{'dns_timeout'};
+
+  my $return;
+
+  eval {
+    $handler->set_handler_alarm( ( $timeout + 0.1 ) * 1000000 ); # 0.1 seconds over that passed to Net::DNS::Resolver
+    if ( $what eq 'query' ) {
+      $return = $self->SUPER::send( @_ );
+    }
+    elsif ( $what eq 'search' ) {
+      $return = $self->SUPER::send( @_ );
+    }
+    elsif ( $what eq 'send' ) {
+      $return = $self->SUPER::send( @_ );
+    }
+    $handler->reset_alarm();
+  };
+
```

Comment:
Tiny window here for `SIGALRM` to be delivered outside of the eval, before the alarm is reset? Probably not a ton we can do about it though?

Reply:
I don't think there is potential for them to go uncaught this way.